🐛 Destination S3: fixed s3 destination field naming for Parquet and Avro formats

andriikorotkov opened the pull request:
## What
The bug is described in detail in task https://github.com/airbytehq/airbyte/issues/5149. Issue seems to be that some fields in the source start with a numeric character.

## How
To avoid an invalid character bug when writing to avro/parquet, the "_" prefix has been added.

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector added to connector index like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
   
</p>
</details>


<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector version bumped like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

etsybaev reviewed (approved):
LGTM, but please make sure all check\tests passed before moving forward

andriikorotkov commented:
/test connector=connectors/destination-s3

andriikorotkov commented:
/test connector=connectors/destination-s3

andriikorotkov commented:
/test connector=connectors/destination-s3

andriikorotkov commented:
/test connector=connectors/destination-s3

tuliren reviewed (changes requested):
Adding the underscore to every identifier seems unnecessary. Can we do this just for names with invalid characters? I don't want all of our users to have to deal with this extra underscore just because facebook has an API that behaves weirdly. Facebook does not deserve this treatment.

andriikorotkov commented:
@tuliren I can make the changes you mentioned above. I corrected this in the way the @sherifnada talked about in task https://github.com/airbytehq/airbyte/issues/5149. Sherif, could you tell me whether to leave the underscore for all identifiers or do it the way Liren suggested?

sherifnada commented:
Liren is correct here IMO we should only do it for streams which start with invalid chars

andriikorotkov commented:
@tuliren, I fixed the name transformer as you mentioned in the comment above. Please take a look at this pull request again.

sherifnada commented on airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/S3NameTransformer.java:
```diff
@@ -37,7 +37,15 @@ protected String applyDefaultCase(String input) {
+  }
+
+  private String checkFirsCharInStreamName(String name) {
+    if (name.substring(0, 1).matches("\\d")) {
```
would it be better if this should instead check if the first char is a valid first char (whitelist approach instead of blacklist)?

andriikorotkov replied:
of course I have updated s3 name transformer.

sherifnada commented on airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/S3NameTransformer.java:
```diff
@@ -0,0 +1,51 @@
+import com.google.common.annotations.VisibleForTesting;
+import io.airbyte.integrations.destination.ExtendedNameTransformer;
+
+@VisibleForTesting
```
is this annotation needed? not seeing a test for this

sherifnada reviewed (approved):
LGTM - could you also add a note in the docs about this behavior, and add a unit test for S3NameTransformer? 

We might also need to update/publish the GCS destination since it relies on the same S3 utilities

tuliren reviewed (approved):
Looks good. Thank you.

Maybe add a unit test?

andriikorotkov commented:
@sherifnada, I fixed all the remarks, updated the documentation and added tests. the problem is that not all jobs are successful. can i merge this pull request without it?

sherifnada commented:
@andriikorotkov yup

andriikorotkov commented:
/test connector=connectors/destination-s3

andriikorotkov commented:
/publish connector=connectors/destination-s3

andriikorotkov commented:
/test connector=connectors/destination-s3

andriikorotkov commented:
/publish connector=connectors/destination-s3
